### ISSUE TITLE
vJoy axis response curves do not work with relative axis remaps

### ISSUE THREAD

**Issue by Jahfry** [WhiteMagic/JoystickGremlin#400]
I'll try to explain, this might be a bit of a mess. 

## Problem:

***Summary:*** adjusting Response Curve on an axis set to Relative mode blocks input from that axis. 

***Workaround?*** Haven't found one yet aside from just getting used to this axis working in the wrong direction as Star Citizen doesn't give an inversion option for mining laser power axis. 

***Details:***

I have a setup (see details below) where if I hold button 3 on my right stick, and twist the right stick, it sends input to vjoy1 as Zrot. Binding that in Star Citizen allows use as a mining throttle. Works.

Setting to Relative mode works perfect. 

I then apply a Response Curve to vjoy1 Zrot. Works. 

But ... if I use the Response Curve "Invert" toggle ... input is no longer seen in game on that vjoy1 Zrot axis. 

Reverting the Invert -or- removing the curve -or- switching vjoy1 Zrot back to "Absolute" (and restarting profile each time) immediately gives input to game. 

Doing a manual invert (dragging control points with mouse instead of using Invert button) doesn't change the story. 

With further testing, **any Response Curve** aside from the default & uninverted setting causes Relative vjoy1 Zrot values to stop being seen in SC. Even just a small lowering of the default uninverted curve. 

Issue happens with Cubic Spline as well as Bezier when inverted. 

*Extra Note:*

* When switching vjoy2 Zrot to Relative (from Absolute), input doesn't seem to register at scale 1.00 ... scale 5.00 is decent ... scale 10.00 is responsive. Just noting in case this is unexpected as I assumed the 1.00 default should give standard level of response. 

### Hardware/OS/Software: 

* VKB SCG Premium + Gunfighter mk2 bases
* Windows 10 Pro 21H2
* vJoy 2.1.9.1
* Joystick Gremlin 13.3
* Star Citizen 3.16.1

### VKB Setup

* Firmware 2.071
* VKB Dev Config 0.92.17
    * Default settings aside from doing a standard calibration

### Vjoy

* vjoy1
    * all axes enabled
    * 32 buttons
    * 0 hats
    * FFB disabled
* vjoy2
    * all axes enabled
    * 96 buttons
    * 2 hats
    * FFB disabled

### Joystick Gremlin

* vjoy1
     * using for virtual mappings
          * currently only virtual mapping is using vjoy1 Zrot as (input2) Zrot axis for Star Citizen mining laser power when right VKB button3 is held pressed
     * placed VKB SCG Premium ministick axes (Xrot/Yrot)
          * left stick ministick to vjoy1 X/Y
          * right stick to vjoy1 Xrot/Yrot
     * Summary of used axes: X/Y (left ministick), Xrot/Yrot (right ministick), Zrot (virtual mining throttle axis). Unused: Z, Slider, Dial
* vjoy2
    * using for combing left and right VKB standard inputs into a single device
         * Left X/Y/Zrot (twist) = vjoy2 X/Y/Z
         * Right X/Y/Zrot (twist) = vjoy Xrot/Yrot/Zrot
         * Left buttons 1 through 27 = vjoy2 1 through 27
         * Right buttons 1 through 27 = vjoy2 41 through 67

**Comment by Jahfry** [comment]
XML of current config attached here, state is with Response Curve on vjoy1 Zrot being inverted and therefor not giving input. Simply reverting the curve to default would then allow the input to work.

[Star Citizen 3.16.1 VKB dual SCG v003 - curve issue.xml.zip](https://github.com/WhiteMagic/JoystickGremlin/files/8276689/Star.Citizen.3.16.1.VKB.dual.SCG.v003.-.curve.issue.xml.zip)

**Comment by Jahfry** [comment]
Possibly related to #363 ?

**Comment by WhiteMagic** [comment]
What are you seeing in Gremlin's input viewer when adding response curves to a relative axis? In my quick tests, adding the response curve behaves as expected. 

As for the scaling values, the code updates the axis value at 100Hz with a step size of `input * (scale / 1000)`, so it will take 10s to go from centre to full deflection when the stick is fully deflected.

**Comment by Jahfry** [comment]
Ok. I think you can close this and mark as "something was screwy". 

The reallllllly strange thing that fixed it? I upgraded Win10 > Win11 after this report (for a completely separate WSL project I'm needing 11 for) and had to strip out and reinstall vJoy and HIDHide to get HIDHide working again. I *didn't* reinstall Gremlin and I didn't **reconfigure** vJoy (but did HIDHide). 

Coming back to this today after all of that and ... inverting the curve fully functions in game for me. Just tested. 

Sorry for the added note (and the time I spent banging my head on testing it 2 days ago lol). Apparently something on my old OS wasn't working right and I have no way of going back to replicate.

**Comment by Jahfry** [comment]
I just did some screenshots showing no input vs  relative vs relative+inverted. 

Summary: input viewer shows the input but in-game when inverted I'm not seeing a change (let me refine that, I do see it jitter between the 50% default state and 49% then back to 50%). 

Shots below:

* No right-stick ztwist ... control example ... ![gremlin01 - nothing pressed](https://user-images.githubusercontent.com/16520816/159094749-2be73aeb-fadc-484a-8d6c-283a0898a219.png)
* right stick full ztwist ... absolute input selected (scale 5.0) ... ![gremlin02 - relative](https://user-images.githubusercontent.com/16520816/159094827-606d18a9-d08a-42db-abeb-cdb536a09a9b.png)
 * same except curve set on vjoy1 inverted with button in UI ... ![gremlin03 - relative inverted](https://user-images.githubusercontent.com/16520816/159094887-d4a08d62-d916-4783-b88e-85a841ffc1f8.png)

Note: in the above 3rd example, if the invert had worked, the game screen would have shown mining throttle at 100% (not the 0%) from the inversion. No amount of twisting and holding either direction would budge it from 49-50% (50% solid if scale 1.00 instead of 5.00). 

### Additional notes:

* If I have the throttle set to 0 in game prior to activating the throttle axis, and axis curve inverted, it jumps to 50% when twisted and then jitters sometimes to 49% (then back to 50%). 
* **important?** If remove the response curve from vjoy1 Zrot and instead apply to the right stick action (I hold button 3 + use ztwist to activate this throttle, so curve is added only to that action and not the primary ztwist action without button3) ... *things work as expected* ... throttle in game is responsive in relative+inverted mode. 

Until that last note I was leaning towards this maybe being a problem in the game itself, not Gremlin. But with that last note above it really seems to be something in Gremlin going strange on the vjoy curve (since it works as a physical stick curve). 

I can handle using that as a workaround, it's no easier/harder to apply the curve in that location. 

Attaching an updated XML to show what did work (contrasted with the earlier .ZIP that did not work) 

[Star Citizen 3.16.1 VKB dual SCG v005 - workaround.zip](https://github.com/WhiteMagic/JoystickGremlin/files/8308051/Star.Citizen.3.16.1.VKB.dual.SCG.v005.-.workaround.zip)

**Comment by WhiteMagic** [comment]
Just to clarify the response curve is on the vJoy device and not on the physical device? If yes then this might be the actual issue as I haven't tried this. Response curves on vJoy axes directly behave slightly different from the normal ones added to physical inputs.

**Comment by Jahfry** [comment]
I've tested both. Summary of results:

1) Response Curve, inverted, on vJoy device = no input

2) Response Curve, inverted, on physical VKB devies = works as expected

**Comment by WhiteMagic** [comment]
Yeah, something is off with how response curves that aren't doing anything affect the relative input.

**Comment by WhiteMagic** [comment]
R14 no longer supports on output vJoy device response curves.
